Source/dataframe.lisp

```diff
@@ -0,0 +1,245 @@
+                                 col
+                                 (position col cols :test #'string=))))
+                      (etypecase col-slice
+                        ((eql t)
+                         (length cols))
+                        ((cons (eql t) %col-index%)
+                         (1+ (norm-index (cdr col-slice))))
+                        ((cons %col-index% (eql t))
+                         (- (length cols) (norm-index (car col-slice))))
+                        ((cons %col-index% %col-index%)
+                         (1+ (- (norm-index (cdr col-slice)) (norm-index (car col-slice)))))
+                        ((vector %col-index%)
+                         (length col-slice)))))
+           (new-rows (make-array row-num :adjustable t))
+           (new-cols (make-array col-num :adjustable t)))
+      (let ((index 0))
+        (map-data-frame-cols df nil col-slice
+                             (lambda (col-index col-name value)
+                               (declare (ignore col-index value))
+                               (setf (elt new-cols index) col-name)
+                               (incf index))))
+      (let ((index 0))
+        (flet ((copy-row (src dst &aux (index 0))
+                 (map-data-frame-cols df src col-slice
+                                      (lambda (col-index col-name value)
+                                        (declare (ignore col-index col-name))
+                                        (setf (elt dst index) value)
+                                        (incf index)))))
+          (map-data-frame-rows df row-slice
+                               (lambda (row-index row)
+                                 (declare (ignore row-index))
+                                 (let ((new-row (make-array col-num :adjustable t)))
+                                   (copy-row row new-row)
+                                   (setf (elt new-rows index) new-row)))))))))
+
+(defgeneric map-data-frame (<data-frame> %row-slice% %col-slice% function)
+  (:method ((df <raw-data-frame>) row-slice col-slice fun)
+    (map-data-frame-rows
+     df row-slice
+     (lambda (row-index row)
+       (map-data-frame-cols
+        df row col-slice
+        (lambda (col-index col-name value)
+          (funcall fun row-index row col-index col-name value)))))))
+
+(defgeneric map-data-frame-rows (<data-frame> %row-slice% function)
+  (:method ((df <raw-data-frame>) row-slice fun)
+    (assert (typep row-slice '%row-slice%) nil '<row-invalid-slice>)
+    (let ((rows (rows df)))
+      (etypecase row-slice
+        ((eql t)
+         (loop for index from 0
+               for row across rows
+               do (funcall fun index row)))
+        ((cons (eql t) %row-index%)
+         (loop for index from 0 upto (cdr row-slice)
+               for row across rows
+               do (funcall fun index row)))
+        ((cons %row-index% (eql t))
+         (loop for index from (car row-slice)
```

Comment:
should be
`
         (loop for index from (car row-slice) below (length (rows df))
`

Reply:
good catch, fixed